src/main/java/fi/aalto/cs/apluscourses/presentation/MainViewModel.java

```diff
@@ -134,16 +133,36 @@ public Options getExerciseFilterOptions() {
     return exerciseFilterOptions;
   }
 
-  private void setInGrading(List<ExerciseGroup> exerciseGroups) {
-    final long id = submittedForGrading.getId();
+  private static void setInGrading(List<ExerciseGroup> exerciseGroups, long id) {
     exerciseGroups
         .stream()
         .map(ExerciseGroup::getExercises)
         .filter(exercise -> exercise.containsKey(id))
         .forEach(exercise -> exercise.get(id).setInGrading(true));
   }
 
-  public void setSubmittedForGrading(Exercise submittedForGrading) {
-    this.submittedForGrading = submittedForGrading;
+  /**
+   * Modifies the 'inGrading' status of the exercise in 'exercisesViewModel' which corresponds to
+   * the given exercise (i.e. has the same ID). The given exercise is not modified. The observers of
+   * 'exercisesViewModel' are notified.
+   */
+  public void setSubmittedForGrading(@NotNull Exercise submittedForGrading) {
+    Exercise previous = inGrading.putIfAbsent(submittedForGrading.getId(), submittedForGrading);
+    if (previous == null) {
+      // This method gets called repeatedly by SubmissionStatusUpdater, so we check whether the
+      // entry was already there or not, and only update the tree if it wasn't there previously.
+      setInGrading(exercisesViewModel.get().getModel(), submittedForGrading.getId());
```

Comment:
perhaps ensure _`nonNull`_ model is returned?

Reply:
It's luckily annotated with `@NotNull`